File: zettelkasten-front/src/api/habits.ts
import { apiClient } from './client';
import { Habit, HabitWithCheckin, HabitLog, HabitStats, CreateHabitParams, UpdateHabitParams, CheckinHabitParams } from '../models/habit';

/**
 * Get all habits for the current user
 */
export async function getHabits(): Promise<Habit[]> {
  const { data } = await apiClient.get<Habit[]>('/habits');
  return data;
}

/**
 * Get habits due today with check-in status
 */
export async function getTodaysHabits(): Promise<HabitWithCheckin[]> {
  const { data } = await apiClient.get<HabitWithCheckin[]>('/habits/today');
  return data;
}

/**
 * Get a single habit by ID
 */
export async function getHabit(id: number): Promise<Habit> {
  const { data } = await apiClient.get<Habit>(`/habits/${id}`);
  return data;
}

/**
 * Get statistics for a specific habit
 */
export async function getHabitStats(id: number): Promise<HabitStats> {
  const { data } = await apiClient.get<HabitStats>(`/habits/${id}/stats`);
  return data;
}

/**
 * Get check-in logs for a specific habit
 */
export async function getHabitLogs(id: number, limit?: number, offset?: number): Promise<HabitLog[]> {
  const params: Record<string, number | undefined> = {};
  if (limit !== undefined) params.limit = limit;
  if (offset !== undefined) params.offset = offset;

  const { data } = await apiClient.get<HabitLog[]>(`/habits/${id}/logs`, { params });
  return data;
}

/**
 * Create a new habit
 */
export async function createHabit(params: CreateHabitParams): Promise<number> {
  const { data } = await apiClient.post<{ id: number }>('/habits', params);
  return data.id;
}

/**
 * Update an existing habit
 */
export async function updateHabit(id: number, params: UpdateHabitParams): Promise<Habit> {
  const { data } = await apiClient.put<Habit>(`/habits/${id}`, params);
  return data;
}

/**
 * Delete a habit
 */
export async function deleteHabit(id: number): Promise<void> {
  await apiClient.delete(`/habits/${id}`);
}

/**
 * Check in to a habit (mark as complete for today)
 */
export async function checkinHabit(id: number, params?: CheckinHabitParams): Promise<number> {
  const { data } = await apiClient.post<{ id: number }>(`/habits/${id}/checkin`, params || {});
  return data.id;
}

/**
 * Undo a habit check-in
 */
export async function undoCheckin(id: number, logId: number): Promise<void> {
  await apiClient.delete(`/habits/${id}/checkin/${logId}`);
}

File: go-backend/handlers/habits.go
package handlers

import (
	"encoding/json"
	"log"
	"net/http"
	"strconv"

	"go-backend/models"
	"go-backend/services"

	"github.com/gorilla/mux"
)

func (s *Handler) GetHabitsRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	habits, err := services.GetHabits(s.GetDB(), userID)
	if err != nil {
		log.Printf("Error getting habits: %v", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(habits)
}

func (s *Handler) GetHabitRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	id, err := strconv.Atoi(mux.Vars(r)["id"])
	if err != nil {
		log.Printf("Invalid habit id param: %v", err)
		http.Error(w, "Invalid id", http.StatusBadRequest)
		return
	}

	habit, err := services.GetHabit(s.GetDB(), userID, id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(habit)
}

func (s *Handler) CreateHabitRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	var params models.CreateHabitParams
	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
		log.Printf("Error decoding create habit request: %v", err)
		http.Error(w, "Invalid request", http.StatusBadRequest)
		return
	}
	id, err := services.CreateHabit(s.GetDB(), userID, params)
	if err != nil {
		log.Printf("Error creating habit: %v", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(map[string]int{"id": id})
}

func (s *Handler) DeleteHabitRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	id, err := strconv.Atoi(mux.Vars(r)["id"])
	if err != nil {
		log.Printf("Invalid habit id param: %v", err)
		http.Error(w, "Invalid id", http.StatusBadRequest)
		return
	}

	err = services.DeleteHabit(s.GetDB(), userID, id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

func (s *Handler) CheckinHabitRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	id, err := strconv.Atoi(mux.Vars(r)["id"])
	if err != nil {
		log.Printf("Invalid habit id param: %v", err)
		http.Error(w, "Invalid id", http.StatusBadRequest)
		return
	}

	var params models.CheckinHabitParams
	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
		log.Printf("Error decoding checkin request: %v", err)
		http.Error(w, "Invalid request", http.StatusBadRequest)
		return
	}

	timezone, _ := s.GetUserTimezone(userID)
	logID, err := services.CheckinHabit(s.GetDB(), userID, id, params, timezone)
	if err != nil {
		if err.Error() == "already checked in today" {
			http.Error(w, err.Error(), http.StatusConflict)
		} else {
			http.Error(w, err.Error(), http.StatusInternalServerError)
		}
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(map[string]int{"id": logID})
}

func (s *Handler) GetTodaysHabitsRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	timezone, _ := s.GetUserTimezone(userID)
	habits, err := services.GetTodaysHabits(s.GetDB(), userID, timezone)
	if err != nil {
		log.Printf("Error getting today's habits: %v", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(habits)
}

// GetHabitLogsRoute retrieves the check-in history for a specific habit
// GET /api/habits/{id}/logs
func (s *Handler) GetHabitLogsRoute(w http.ResponseWriter, r *http.Request) {
	// TODO: Implement this handler
	http.Error(w, "Not implemented", http.StatusNotImplemented)
}

// GetHabitStatsRoute retrieves statistics for a specific habit
// GET /api/habits/{id}/stats
func (s *Handler) GetHabitStatsRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	id, err := strconv.Atoi(mux.Vars(r)["id"])
	if err != nil {
		log.Printf("Invalid habit id param: %v", err)
		http.Error(w, "Invalid id", http.StatusBadRequest)
		return
	}

	// Verify habit exists and belongs to user
	_, err = services.GetHabit(s.GetDB(), userID, id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	timezone, _ := s.GetUserTimezone(userID)
	stats, err := services.CalculateHabitStats(s.GetDB(), userID, id, timezone)
	if err != nil {
		log.Printf("Error calculating habit stats: %v", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(stats)
}

// UndoCheckinRoute deletes a habit check-in log
// DELETE /api/habits/{id}/checkin/{logId}
func (s *Handler) UndoCheckinRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	vars := mux.Vars(r)

	habitID, err := strconv.Atoi(vars["id"])
	if err != nil {
		log.Printf("Invalid habit id param: %v", err)
		http.Error(w, "Invalid habit id", http.StatusBadRequest)
		return
	}

	logID, err := strconv.Atoi(vars["logId"])
	if err != nil {
		log.Printf("Invalid log id param: %v", err)
		http.Error(w, "Invalid log id", http.StatusBadRequest)
		return
	}

	err = services.DeleteHabitLog(s.GetDB(), userID, habitID, logID)
	if err != nil {
		if err.Error() == "habit log not found" {
			http.Error(w, err.Error(), http.StatusNotFound)
		} else {
			http.Error(w, err.Error(), http.StatusInternalServerError)
		}
		return
	}

	w.WriteHeader(http.StatusNoContent)
}
